In [12]:
import nltk

import re
import numpy as np
import pandas as pd
from nltk import RegexpTokenizer
from sklearn.model_selection import train_test_split
from nltk.corpus import stopwords
from nltk import word_tokenize, wordpunct_tokenize
from sklearn.naive_bayes import MultinomialNB
#from sklearn.linear_model import LogisticRegression
#from sklearn.dummy import DummyClassifier
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics import classification_report



In [2]:
# чтобы посмотреть на таблицу
df = pd.read_csv('songdata.csv')
df.head()

artist                   song                                        link  \
0   ABBA  Ahe's My Kind Of Girl  /a/abba/ahes+my+kind+of+girl_20598417.html   
1   ABBA       Andante, Andante       /a/abba/andante+andante_20002708.html   
2   ABBA         As Good As New        /a/abba/as+good+as+new_20003033.html   
3   ABBA                   Bang                  /a/abba/bang_20598415.html   
4   ABBA       Bang-A-Boomerang      /a/abba/bang+a+boomerang_20002668.html   

                                                text  
0  Look at her face, it's a wonderful face  \nAnd...  
1  Take it easy with me, please  \nTouch me gentl...  
2  I'll never know why I had to go  \nWhy I had t...  
3  Making somebody happy is a question of give an...  
4  Making somebody happy is a question of give an...

In [3]:
songsdata = pd.read_csv('songdata.csv', sep=',')
artists = songsdata.groupby(['artist']).size().sort_values(ascending=False)
artists

artist
Donna Summer                                    191
Gordon Lightfoot                                189
Bob Dylan                                       188
George Strait                                   188
Reba Mcentire                                   187
Loretta Lynn                                    187
Alabama                                         187
Cher                                            187
Dean Martin                                     186
Chaka Khan                                      186
Hank Williams Jr.                               185
Neil Young                                      185
Nazareth                                        184
Cliff Richard                                   184
Indigo Girls                                    184
America                                         184
Johnny Cash                                     183
Kiss                                            183
Chris Rea                                       182
Bon J

Кажется, датасет не очень сбалансирован, поэтому я лучше возьму исполнителей с топом-количеством песен для чистоты эксперимента. Например, пусть это будут исполнители с 188 песнями и выше.

In [4]:
artists_new = artists[:6]
bands = songsdata.loc[(songsdata['artist'] == 'Donna Summer') | (songsdata['artist'] == 'Gordon Lightfoot') |
                  (songsdata['artist'] == 'Bob Dylan') | (songsdata['artist'] == 'George Strait')
            ]
artists_new
bands

artist                                   song  \
1576          Bob Dylan                        4Th Time Around   
1577          Bob Dylan                       A Satisfied Mind   
1578          Bob Dylan                     Ain't Gonna Grieve   
1579          Bob Dylan                           All Over You   
1580          Bob Dylan                 Ballad Of Donald White   
1581          Bob Dylan                 Ballad Of Hollis Brown   
1582          Bob Dylan                             Belle Isle   
1583          Bob Dylan               Beyond Here Lies Nothin'   
1584          Bob Dylan                     Beyond The Horizon   
1585          Bob Dylan                       Black Crow Blues   
1586          Bob Dylan                    Blind Willie Mctell   
1587          Bob Dylan                    Blowin' In The Wind   
1588          Bob Dylan                      Bob Dylan's Blues   
1589          Bob Dylan                      Bob Dylan's Dream   
1590          Bob Dylan            Bob Dylan's New Orleans Rag   
1591          Bob Dylan                      Broke Down Engine   
1592          Bob Dylan                             California   
1593          Bob Dylan  Can You Please Crawl Out Your Window?   
1594          Bob Dylan             Can't Help Falling In Love   
1595          Bob Dylan                                Catfish   
1596          Bob Dylan                      Chimes Of Freedom   
1597          Bob Dylan                       Christmas Island   
1598          Bob Dylan                          Clean-Cut Kid   
1599          Bob Dylan                           Clothes Line   
1600          Bob Dylan                          Copper Kettle   
1601          Bob Dylan                            Country Pie   
1602          Bob Dylan                            Cry A While   
1603          Bob Dylan                              Dark Eyes   
1604          Bob Dylan                          Dear Landlord   
1605          Bob Dylan                                 Denise   
...                 ...                                    ...   
34331  Gordon Lightfoot                   Race Among The Ruins   
34332  Gordon Lightfoot                          Rainbow Trout   
34333  Gordon Lightfoot                       Rainy Day People   
34334  Gordon Lightfoot                             Red Velvet   
34335  Gordon Lightfoot                           Redwood Hill   
34336  Gordon Lightfoot                            Remember Me   
34337  Gordon Lightfoot                               Restless   
34338  Gordon Lightfoot                          Ringneck Loon   
34339  Gordon Lightfoot                                Romance   
34340  Gordon Lightfoot                      Same Old Loverman   
34341  Gordon Lightfoot                       Saturday Clothes   
34342  Gordon Lightfoot                   Second Cup Of Coffee   
34343  Gordon Lightfoot                                Shadows   
34344  Gordon Lightfoot                     She's Not The Same   
34345  Gordon Lightfoot                          Slide On Over   
34346  Gordon Lightfoot                                 Softly   
34347  Gordon Lightfoot                  Someone To Believe In   
34348  Gordon Lightfoot                 Something Very Special   
34349  Gordon Lightfoot                       Sometimes I Wish   
34350  Gordon Lightfoot                       Somewhere U.S.A.   
34351  Gordon Lightfoot                             Stay Loose   
34352  Gordon Lightfoot                       Steel Rail Blues   
34353  Gordon Lightfoot                       Stone Cold Sober   
34354  Gordon Lightfoot                       Summertime Dream   
34355  Gordon Lightfoot                                Sundown   
34356  Gordon Lightfoot                  Talking In Your Sleep   
34357  Gordon Lightfoot             Thank You For The Promises   
34358  Gordon Lightfoot                That Same Old Obsession   
34359  Gordon Lightfoot    The First Time Ever I Saw Your

In [5]:
# Токенизируем
def tokenizers(text):    
    tokenizer = RegexpTokenizer(r'\w+')
    return tokenizer.tokenize(text)

In [6]:
#  Для начала делим всю нашу выборку в соотношении 20:80
text_train, text_test, artist_train, artist_test = train_test_split(bands['text'], bands['artist'], test_size=0.2)

In [7]:
# Добавим cтоп-слова
stopwords = stopwords.words('english')

In [13]:
# Векторизация и Байес
res = Pipeline([('bow', CountVectorizer(analyzer=tokenizers, stop_words=stopwords)),
                ('classifier', MultinomialNB())])
res.fit(text_train, artist_train)
artist_pred = res.predict(text_test)
print(classification_report(artist_test, artist_pred))

                  precision    recall  f1-score   support

       Bob Dylan       0.62      0.44      0.52        34
    Donna Summer       0.49      0.61      0.54        31
   George Strait       0.64      0.70      0.67        40
Gordon Lightfoot       0.64      0.62      0.63        47

     avg / total       0.61      0.60      0.60       152



Хуже всего определяется Donna Summer, но её песен больше всего...